### PR TITLE
fix: resolve dtype, decoding, and padding issues for safety evaluation

### LINKED ISSUE #4: [BUGs] Deterministic decoding, correct padding for causal LMs, and probability-based scoring
## Summary

The current evaluation script has several issues that impact **numerical accuracy**, **determinism**, and **batch safety** for safety evaluation:

1. **Hard-coded `torch_dtype=torch.float16`** — LLaMA 3.1 models (and many fine-tunes) are trained in **bfloat16**, not float16, but this model loads in fp16 here. 
2. **Non-deterministic decoding** — `temperature=0.2` and `do_sample=True` cause run-to-run variability.
3. **Right padding for causal LM** — tokenizer does not set `padding_side="left"`, which can cause misalignment in batched inference.
(optinal) 4. **Fragile label extraction** — regex parsing ignores token probabilities; borderline cases can be misclassified.

---

## 1. Correct dtype when load model

**Current**:

```python
AutoModelForCausalLM.from_pretrained(
    model_id,
    torch_dtype=torch.float16,
    device_map="auto"
)
```

**Problem**:

* LLaMA 3.1 is in bfloat16 and your fine-tunes are typically trained in **bfloat16**.
* Using `float16` for inference can cause loss of precision in activations and logits, which is especially important for classification-style safety eval.
* bf16 also has a wider dynamic range than fp16, reducing overflow/underflow risk.

**Recommendation**:

* Detect and default to `torch.bfloat16` for LLaMA 3.1 and compatible fine-tunes.

Example:

```python
import torch

device = (
    "cuda" if torch.cuda.is_available()
    else ("mps" if getattr(torch.backends, "mps", None) and torch.backends.mps.is_available() else "cpu")
)

def pick_dtype():
    if device == "cuda":
        if torch.cuda.is_bf16_supported():
            return torch.bfloat16
        return torch.float16
    if device == "mps":
        return torch.bfloat16 if getattr(torch.backends, "mps", None) and hasattr(torch.backends.mps, "is_bf16_supported") and torch.backends.mps.is_bf16_supported() else torch.float16
    return torch.float32

dtype = pick_dtype()
kwargs = dict(torch_dtype=dtype)
if device != "cpu":
    kwargs["device_map"] = "auto"

self.model = AutoModelForCausalLM.from_pretrained(model_id, **kwargs)
```

---

## 2. Deterministic decoding for evaluation

**Current**:

```python
outputs = self.model.generate(
    **inputs,
    max_new_tokens=max_length,
    temperature=0.2,
    top_p=1,
    do_sample=True
)
```

**Problem**: Sampling introduces randomness → same input may yield different outputs.

**Fix**:

```python
outputs = self.model.generate(
    **inputs,
    max_new_tokens=max_length,
    do_sample=False,
    temperature=0.0,
    top_p=None
)
```

---

## 3. Correct padding for causal LMs

**Current**: No `padding_side` set → defaults to `"right"`.

**Fix**:

```python
if self.tokenizer.pad_token is None:
    self.tokenizer.pad_token = self.tokenizer.eos_token
self.tokenizer.padding_side = "left"
```

or

```python
tokenizer = AutoTokenizer.from_pretrained(model_id, padding_side="left")
```

> Left padding ensures correct context alignment in batch generation for causal LMs.

---

## 4. Probability-based scoring instead of regex parsing

**Current**: Regex searches for `"label": "safe"` in free-form text.

**Problems**:

* Brittle to output formatting changes.
* No confidence score.
* Can misclassify borderline cases.

**Fix**: Compute probabilities for `safe` vs `unsafe` tokens directly from logits:

```python
import torch.nn.functional as F

def score_safe_unsafe(self, prompt, candidates=("safe", "unsafe")):
    enc = self.tokenizer(prompt, return_tensors="pt", padding=True, truncation=True)
    enc = {k: v.to(self.model.device) for k, v in enc.items()}
    with torch.no_grad():
        logits = self.model(**enc).logits

    pad_id = self.tokenizer.pad_token_id
    last_idx = (enc["input_ids"] != pad_id).sum(dim=1) - 1
    last_logits = logits[0, last_idx[0], :]
    probs = F.softmax(last_logits, dim=-1)

    cand_ids = {c: self.tokenizer.encode(c, add_special_tokens=False)[0] for c in candidates}
    cand_probs = {c: float(probs[cid]) for c, cid in cand_ids.items()}
    total = sum(cand_probs.values())
    cand_probs = {k: v / total for k, v in cand_probs.items()}
    cand_probs["pred"] = max(cand_probs, key=cand_probs.get)
    return cand_probs
```

---

## Impact

* **Numerically correct** evaluation for LLaMA 3.1 fine-tunes (bf16 consistency).
* **Reproducible** results (no sampling).
* **Batch-safe** generation.
* **More accurate** predictions with probabilities and confidence scores.

### PR DESCRIPTION
**Pull request by Dhanush010**
## Summary
   Fixes 4 critical bugs in the evaluation script that affected numerical accuracy, determinism, and batch safety.

   ## Changes
   1. **Dynamic dtype detection** - Uses bfloat16 when supported (correct for LLaMA 3.1 models)
   2. **Deterministic decoding** - Removed sampling parameters for reproducible results
   3. **Left padding** - Set `padding_side="left"` for correct batch alignment in causal LMs
   4. **Probability scoring** - Added `score_safe_unsafe()` method as alternative to regex parsing

   ## Impact
   - ✅ Reproducible evaluation results
   - ✅ Correct precision for LLaMA 3.1 models
   - ✅ Batch-safe generation
   - ✅ More accurate predictions with confidence scores

   ## Verification
   - Code inspection completed
   - All fixes verified statically
   - No linting errors

   Fixes #4

### CODE REVIEW

File: inference.py [hunk 1]
```diff
@@ -103,14 +129,44 @@ def generate_text(self, prompt: str, max_length: Optional[int] = None) -> str:
             outputs = self.model.generate(
                 **inputs,
                 max_new_tokens=max_length,
-                temperature=0.2,
-                top_p=1,
-                do_sample=True,
+                do_sample=False,
```
[the remark this answers is not in the document]
Comment (youngwanlim) [reply inference.py]: Thanks for the updates. Could you set the default temperature to 0.2? The score is lower when using 0.0 compared to 0.2.

[the remark this answers is not in the document]
Comment (Dhanush010) [reply inference.py]: Thanks for the feedback! I've updated the default temperature to 0.2 as requested. The `generate_text()` method now defaults to `temperature=0.2` for better scoring accuracy.

The change has been committed and pushed.

File: inference.py [hunk 2]
```diff
@@ -100,14 +132,23 @@ def generate_text(self, prompt: str, max_length: Optional[int] = None) -> str:
         inputs = {key: value.to(self.model.device) for key, value in inputs.items()}
 
         with torch.no_grad():
-            outputs = self.model.generate(
-                **inputs,
-                max_new_tokens=max_length,
-                temperature=0.2,
-                top_p=1,
```
[the remark this answers is not in the document]
Comment (Dhanush010) [reply inference.py]: Thanks for asking! I removed top_p when initially implementing deterministic decoding (as part of issue https://github.com/Roblox/RobloxGuard-1.0/issues/4's requirement to remove sampling parameters). However, since we're now using temperature sampling with the default temperature=0.2, it makes sense to restore top_p=1 to match the original code behavior.

I've added it back in the latest commit.

**Review by youngwanlim — APPROVED** [review]
thanks for your contribution.  👍

**Comment by vai-minzhou** [comment]
😁
